File: Sources/Commands/Remove.swift
//
//  Remove.swift
//  
//
//  Created by Owen Cruz-Abrams on 9/21/23.
//

import Foundation
import ArgumentParser

struct RemoveOptions: ParsableArguments {

    @Flag(name: [.long, .customShort("o")], help: "Produce no output while removing aliases.")
    var noOutput: Bool = false

    @Flag(name: [.long], help: "Reload the shell after removing the alias.")
    var reload: Bool = false

    @Argument(help: "The aliases to remove from the \u{001b}[1m.ally\u{001b}[0m file.")
    var alias: [String]

    @Flag(help: "Scan the \u{001b}[1m.zshrc\u{001b}[0m file for the aliases AS WELL AS the \u{001b}[1m.ally\u{001b}[0m file.")
    var zshrc: Bool = false

}

extension Ally {
    struct Remove: ParsableCommand {
        static var configuration = CommandConfiguration(
            commandName: "remove",
            abstract: "Remove an alias from the Ally/ZSH config."
        )
        @OptionGroup var options: RemoveOptions

        mutating func run() throws {

            // Now, find the line that says alias X
            for alias in options.alias {
                try scanAndRemoveIfNeeded(alias: alias, fileLocation: Ally.dotFileLocation, conditionalOutput: !options.noOutput)
                if options.zshrc {
                    try scanAndRemoveIfNeeded(
                        alias: alias,
                        fileLocation: Ally.zshrcFileLocation,
                        conditionalOutput: !options.noOutput
                    )
                }
            }
        }
    }
}

func scanAndRemoveIfNeeded(alias: String, fileLocation: URL, conditionalOutput: Bool) throws {
    var dotAllyContents = try String(contentsOf: fileLocation, encoding: .utf8)
    var lines = dotAllyContents.split(separator: "\n").map { String($0) }
    let finder = "alias \(alias)"
    var indexesToRemove: [Int] = []
    
    for (index, line) in lines.enumerated() {
        if line.localizedCaseInsensitiveContains(finder) {
            indexesToRemove.append(index)
            
            // Check if this is a multi-line alias by looking for unclosed quotes
            let parts = line.split(separator: "=", maxSplits: 1).map { String($0) }
            if parts.count >= 2 {
                let commandPart = parts[1]
                // If the command starts with a quote but doesn't end with one, it's multi-line
                if commandPart.hasPrefix("\"") && !commandPart.dropFirst().hasSuffix("\"") {
                    // Continue collecting lines until we find the closing quote
                    var currentIndex = index + 1
                    while currentIndex < lines.count {
                        indexesToRemove.append(currentIndex)
                        if lines[currentIndex].hasSuffix("\"") {
                            break
                        }
                        currentIndex += 1
                    }
                }
            }
            
            // Remove all lines above the index that are #s as well
            for reverseIndex in stride(from: index - 1, through: 0, by: -1) {
                let prevLine = lines[reverseIndex]
                if prevLine.trimmingCharacters(in: .whitespacesAndNewlines).starts(with: "#") && !prevLine.isEmpty {
                    indexesToRemove.append(reverseIndex)
                } else if !prevLine.isEmpty {
                    break
                }
            }
            
            if conditionalOutput {
                print("Alias removed from \(fileLocation.lastPathComponent): ", terminator: "")
                boldPrint(alias)
            }
            break // Only remove the first match
        }
    }
    
    // Remove lines in reverse order to maintain correct indices
    for index in indexesToRemove.sorted(by: >) {
        lines.remove(at: index)
    }
    
    dotAllyContents = lines.joined(separator: "\n") + "\n"
    try dotAllyContents.write(to: fileLocation, atomically: true, encoding: .utf8)
    if conditionalOutput {
        print("Dot Ally file resaved. Alias ", terminator: "")
        boldPrint(alias, terminator: " ")
        print("No longer available.")
        print("[WARNING] Terminal may require reload.")
    }
}

File: src/commands/remove.py
import argparse
import os

class RemoveOptions:
    def __init__(self):
        parser = argparse.ArgumentParser(description="Remove an alias from the Ally/ZSH config.")
        parser.add_argument('-o', '--no-output', action='store_true', help="Produce no output while removing aliases.")
        parser.add_argument('--reload', action='store_true', help="Reload the shell after removing the alias.")
        parser.add_argument('alias', nargs='+', help="The aliases to remove from the .ally file.")
        parser.add_argument('--zshrc', action='store_true', help="Scan the .zshrc file for the aliases AS WELL AS the .ally file.")
        self.args = parser.parse_args()

class AllyRemove:
    def __init__(self):
        self.options = RemoveOptions().args

    def run(self):
        for alias in self.options.alias:
            self.scan_and_remove_if_needed(alias, self.dot_file_location(), not self.options.no_output)
            if self.options.zshrc:
                self.scan_and_remove_if_needed(alias, self.zshrc_file_location(), not self.options.no_output)

    def scan_and_remove_if_needed(self, alias, file_location, conditional_output):
        with open(file_location, 'r') as file:
            lines = file.readlines()

        finder = f"alias {alias}"
        new_lines = []
        skip_next = False

        for line in lines:
            if skip_next:
                skip_next = False
                continue
            if finder in line:
                if conditional_output:
                    print(f"Alias removed from {os.path.basename(file_location)}: {alias}")
                continue
            if line.strip().startswith("#") and new_lines and new_lines[-1].strip().startswith(finder):
                skip_next = True
                continue
            new_lines.append(line)

        with open(file_location, 'w') as file:
            file.writelines(new_lines)

        if conditional_output:
            print(f"Dot Ally file resaved. Alias {alias} no longer available.")
            print("[WARNING] Terminal may require reload.")

    @staticmethod
    def dot_file_location():
        return os.path.expanduser("~/.ally")

    @staticmethod
    def zshrc_file_location():
        return os.path.expanduser("~/.zshrc")

if __name__ == "__main__":
    AllyRemove().run()
